DAOS-16953 tests: fix btree parameter parsing

grom72 opened the pull request:
In btree.sh the -m and -t parameters were not applied correctly because of the preceding positional argument containing spaces but not wrapped in quotes

In btree.c an invalid parameter sequence was not detected due to of the partially manual parsing process.

btr_test_state introduced to provide global parameters to the test (argc/argv) and avoid static variables for that purpose.

Additionally:
instead of `-t` and `-m` parameters new parameters has been proposed:
* `-R[d]` for class registration, optionally with dynamic root
* `-M[p]` for memory registration, optionally with indication to use persistent memory

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grom72/daos/79)
<!-- Reviewable:end -->

grom72 committed:
DAOS-16953 tests: fix btree parameter parsing

In btree.sh the -m and -t parameters were not applied correctly
because of the preceding positional argument containing spaces but
not wrapped in quotes

In btree.c an invalid parameter sequence was not detected
due to of the partially manual parsing process.

btr_test_state introduced to provide global parameters to test
(argc/argv) and avoid static variables for that purpose.

Co-authored-by: Jan Michalski <jan-marian.michalski@hpe.com>
Co-authored-by: Tomasz Gromadzki <tomasz.gromadzki@hpe.com>

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@hpe.com>

janekmi reviewed (commented):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: all files reviewed, 9 unresolved discussions (waiting on @grom72)

___
*[a discussion](https://reviewable.io/reviews/grom72/daos/79#-OIGqjtG9or24tyffEnp:-OIGqjtG9or24tyffEnq:b-spz55i) (no related file):*
Nice commit message. :thumbsup: 
___
*[`src/common/tests/btree.sh` line 34 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKhXJl29BGlZROsBGd:-OIKhXJl29BGlZROsBGe:undefined) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.sh#L34)):*
> ```Shell
>         dyn       Run with dynamic root
>         ukey      Use integer keys
>         emb       Use integer keys and embedded value
> ```





_Suggestion:_
```Shell
Use embedded value
```
___
*[`src/common/tests/btree.c` line 920 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKpYyh6mbQ8Y6H6YD_:-OIKpYyh6mbQ8Y6H6YDa:b-tm3w54) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L920)):*
> ```C
> 		D_ASSERT(rc == 0);
> 		if (rc)
> 			fail_msg("Cannot setup pmem: %d\n", rc);
> ```

Either-or.



_Suggestion:_
```C
		if (rc) {
			fail_msg("Cannot setup pmem: %d\n", rc);
		}
```
___
*[`src/common/tests/btree.c` line 926 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKpjCgCRB3YVKpRYV9:-OIKpjCgCRB3YVKpRYVA:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L926)):*
> ```C
> 		D_ASSERT(rc == 0);
> 		if (rc)
> 			fail_msg("Cannot setup vmem: %d\n", rc);
> ```

.



_Suggestion:_
```C
		if (rc) {
			fail_msg("Cannot setup vmem: %d\n", rc);
		}
```
___
*[`src/common/tests/btree.c` line 938 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKnroYBm2MpTVM6COw:-OIKnroYBm2MpTVM6COx:b-96n10y) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L938)):*
> ```C
> /**
>  * Execute test based on the given sequence of steps defined by
>  * the test_state->argc and test_state->argv
> ```

It is nice to have a bit of documentation here. But I do not like to put all the technical details in the comment. It makes it hard to maintain.



_Suggestion:_
```C
 * Execute test based on the given sequence of steps.
```
___
*[`src/common/tests/btree.c` line 943 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKp27TEQ9y9JEj9CNS:-OIKp27TEQ9y9JEj9CNT:b-bgvzqz) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L943)):*
> ```C
>  *    otherwise it will be ignored
>  * -m parameter shall be provided before any other short parameters,
>  *    otherwise it will be ignored
> ```

Kudos for pointing it out.



_Code quote:_
```C
 * -t parameter shall be provided as the first parameter in the sequence,
 *    otherwise it will be ignored
 * -m parameter shall be provided before any other short parameters,
 *    otherwise it will be ignored
```
___
*[`src/common/tests/btree.c` line 959 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKnQ9xBgfPELLIH3JP:-OIKnQ9yB98kr610LprA:bce1jma) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L959)):*
> ```C
> 		switch (opt) {
> 		case 'S':
> 			/* General options, not part of the test sequence */
> ```

I do not get this bit. Maybe just drop it?



_Suggestion:_
```C
not part of the test sequence
```
___
*[`src/common/tests/btree.c` line 970 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKlNnp5F8XmDv3KzUS:-OIKlNnp5F8XmDv3KzUT:b-p5i32j) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L970)):*
> ```C
> 		case 'C':
> 			if (ik_utx == NULL)
> 				ik_btr_init_mem(false, dynamic_flag);
> ```

I would prefer to have two switches with optional arguments:

- register class
  - with an option to make it a dynamic root
- initialize memory
  - with an option to make it PMEM

Two benefits to my proposal would be:

1. You would see that registering a class and memory initialization are separate.
2. You would no have weird "leaks" between `-m` and `-C` options just to initialize the memory on time. The memory would be initialized in one place.

Please consider. 



_Code quote:_
```C
		case 't':
			D_PRINT("Using dynamic tree order\n");
			dynamic_flag = BTR_FEAT_DYNAMIC_ROOT;
			break;
		case 'm':
			ik_btr_init_mem(true, dynamic_flag);
			break;
		case 'C':
			if (ik_utx == NULL)
				ik_btr_init_mem(false, dynamic_flag);
```
___
*[`src/common/tests/btree.c` line 1069 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKj5sF2bYPFVJxpwxo:-OIKj5sG9jdM_Urkth06:undefined) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1069)):*
> ```C
> 	}
> 	if (argc != optind) {
> 		print_message("Invalid parameters provided.\n");
> ```





_Suggestion:_
```C
print_message("Unknown parameter: %s\n", argv[optind]);
```
___
*[`src/common/tests/btree.c` line 1073 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKjJH58mNrwrH803UC:-OIKjJH58mNrwrH803UD:bymb0nw) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1073)):*
> ```C
> 	}
> 	if (test_name == NULL)
> 		test_name = "Btree testing tool";
> ```

You might already noticed I am for having braces no matter what.



_Suggestion:_
```C
	if (test_name == NULL) {
		test_name = "Btree testing tool";
	}
```
___
*[`src/common/tests/btree.c` line 1082 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKlHoh2R8zwzXIrZI7:-OIKlHoh2R8zwzXIrZI8:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1082)):*
> ```C
> 	daos_debug_fini();
> 	if (ik_utx)
> 		rc += utest_utx_destroy(ik_utx);
> ```

.



_Suggestion:_
```C
	if (ik_utx) {
		rc += utest_utx_destroy(ik_utx);
	}
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: 1 of 2 files reviewed, 9 unresolved discussions (waiting on @janekmi)

___
*[`src/common/tests/btree.c` line 920 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKpYyh6mbQ8Y6H6YD_:-OIKwhrJ8f5QBDFGuueT:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L920)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Either-or.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 926 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKpjCgCRB3YVKpRYV9:-OIKwhO58aS2wuxMyDTg:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L926)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 938 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKnroYBm2MpTVM6COw:-OIKwOY-AmU9swiZF8nn:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L938)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

It is nice to have a bit of documentation here. But I do not like to put all the technical details in the comment. It makes it hard to maintain.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 959 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKnQ9xBgfPELLIH3JP:-OIKwMUz3ynFq_s6YU3z:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L959)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I do not get this bit. Maybe just drop it?
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 1073 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKjJH58mNrwrH803UC:-OIKwKPe4daeVTJp3KJT:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1073)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

You might already noticed I am for having braces no matter what.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 1082 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKlHoh2R8zwzXIrZI7:-OIKwJNS3toT8FSweN9l:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1082)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.sh` line 34 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKhXJl29BGlZROsBGd:-OIL06La2RxkjyusbeFB:b-hqkn3w) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.sh#L34)):*
> ```Shell
>         dyn       Run with dynamic root
>         ukey      Use integer keys
>         emb       Use integer keys and embedded value
> ```

`emb` also includes `ukey` option
___
*[`src/common/tests/btree.c` line 1069 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKj5sF2bYPFVJxpwxo:-OIKw0Pc3HfziTNOGIkh:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1069)):*
> ```C
> 	}
> 	if (argc != optind) {
> 		print_message("Invalid parameters provided.\n");
> ```

Done.


<!-- Sent from Reviewable.io -->

grom72 committed:
Introduce new parameters for memory and class reg.

Introduce new parameters for explicity class and memory
registration. Both parameters must be provided before
`-C` parameter.

Syntax is as follow:
* -M[p] for memory registration. The `p` value inidcates
  the usage of persisten memory
* -R[d] for class registration. The `d` value indicates
  the usage of `BTR_FEAT_DIRECT_KEY` flag - root is
  dynamically sized up to tree order.

Parameters `-t` and `-d` are no longer valid.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@hpe.com>

grom72 reviewed (commented):
> Reviewed 8 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: 8 of 9 files reviewed, 9 unresolved discussions (waiting on @janekmi)

___
*[`src/common/tests/btree.c` line 970 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKlNnp5F8XmDv3KzUS:-OILIlTb-CDjujaMx_By:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L970)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I would prefer to have two switches with optional arguments:

- register class
  - with an option to make it a dynamic root
- initialize memory
  - with an option to make it PMEM

Two benefits to my proposal would be:

1. You would see that registering a class and memory initialization are separate.
2. You would no have weird "leaks" between `-m` and `-C` options just to initialize the memory on time. The memory would be initialized in one place.

Please consider.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.sh` line 34 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKhXJl29BGlZROsBGd:-OILFxNe8n1s3H7bC5P-:bb6un1j) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.sh#L34)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

`emb` also includes `ukey` option
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: all files reviewed, 9 unresolved discussions (waiting on @janekmi)



<!-- Sent from Reviewable.io -->

janekmi reviewed (commented):
> Reviewed 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: all files reviewed, 12 unresolved discussions (waiting on @grom72)

___
*[`src/common/tests/btree.c` line 1073 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKjJH58mNrwrH803UC:-OILWnBm0-Fw2FCnn5B7:b-83og3x) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1073)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Nope.
___
*[`src/common/tests/btree.sh` line 34 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKhXJl29BGlZROsBGd:-OILWqbH2-PkNHfiQ_ER:b-h3jgr6) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.sh#L34)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Please do it in the first commit.
___
*[`src/common/tests/btree.c` line 892 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILKy_6-3xZqqghgIo-:-OILKy_6-3xZqqghgIo0:undefined) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L892)):*
> ```C
>     {"start-test", required_argument, NULL, 'S'},
>     {"reg_class", optional_argument, NULL, 'R'},
>     {"reg_memory", optional_argument, NULL, 'M'},
> ```





_Suggestion:_
```C
    {"reg-class", optional_argument, NULL, 'R'},
    {"reg-memory", optional_argument, NULL, 'M'},
```
___
*[`src/common/tests/btree.c` line 916 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILVYeN7ewWIvm3SS0j:-OILVYeO3iVyeWEhJFd8:b-1vjw34) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L916)):*
> ```C
> 	int rc = 0;
> 	if (pmem_flag && *pmem_flag != 'p') {
> 		fail_msg("Invalid value of -M parameter: %c\n", *pmem_flag);
> ```

IMHO it is a string not a character.



_Suggestion:_
```C
fail_msg("Invalid value of -M parameter: %s\n", pmem_flag);
```
___
*[`src/common/tests/btree.c` line 940 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILVn0A4BXfmLE6Jf_p:-OILVn0A4BXfmLE6Jf_q:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L940)):*
> ```C
> {
> 	if (dynamic_flag && *dynamic_flag != 'd') {
> 		fail_msg("Invalid value of -R parameter: %c\n", *dynamic_flag);
> ```

.



_Suggestion:_
```C
fail_msg("Invalid value of -R parameter: %s\n", dynamic_flag);
```
___
*[`src/common/tests/btree.c` line 1073 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILWXzwAbnQBCXlUISI:-OILWXzwAbnQBCXlUISJ:bomti6h) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L1073)):*
> ```C
> 			/* unknown parameter */
> 			break;
> 		}
> ```

It doesn't look right. Are you sure you need this despite the `+` at the beginning of the options string?
And how it was not necessary in the previous commit?



_Code quote:_
```C
		if (opt == '?') {
			/* unknown parameter */
			break;
		}
```
___
*[`src/common/tests/btree.sh` line 114 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILX7NM03SnFNT2qZGK:-OILX7NM03SnFNT2qZGL:byf00ys) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L114)):*
> ```Shell
>         eval "${VCMD}" "$BTR"                       \
>         --start-test "'btree functional ${test_conf_pre} ${test_conf} iterate=${IDIR}'" \
>         -R"${DYN}" -M"${PMEM}" -C "${UINT}${IPL}o:$ORDER" \
> ```

It would be nice to either separate value from option with a space for all the options or none of them.



_Code quote:_
```Shell
      -R"${DYN}" -M"${PMEM}" -C "${UINT}${IPL}o:$ORDER" \
```
___
*[`src/common/tests/btree.sh` line 134 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILXKBq9bg7qFGHJKaI:-OILXKBq9bg7qFGHJKaJ:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L134)):*
> ```Shell
>         eval "${VCMD}" "$BTR" \
>         --start-test "'btree batch operations ${test_conf_pre} ${test_conf}'" \
>         -R"${DYN}" -M"${PMEM}" -C "${UINT}${IPL}o:$ORDER" \
> ```

.
___
*[`src/common/tests/btree.sh` line 143 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILXLYD3DQjFCBfoRTL:-OILXLYD3DQjFCBfoRTM:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L143)):*
> ```Shell
>         eval "${VCMD}" "$BTR" \
>         --start-test "'btree drain ${test_conf_pre} ${test_conf}'" \
>         -R"${DYN}" -M"${PMEM}" -C "${UINT}${IPL}o:$ORDER" \
> ```

.
___
*[`src/common/tests/btree.sh` line 150 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILXMHFAESyEpE8BAe8:-OILXMHFAESyEpE8BAe9:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L150)):*
> ```Shell
>         eval "${VCMD}" "$BTR" \
>         --start-test "'btree performance ${test_conf_pre} ${test_conf}'" \
>         -R"${DYN}" -M"${PMEM}" -C "${UINT}${IPL}o:$ORDER" \
> ```

.
___
*[`src/common/tests/btree.c` line 1061 at r2](https://reviewable.io/reviews/grom72/daos/79#-OILUxRSCtZZSNrPs_Vd:-OILUxRSCtZZSNrPs_Ve:undefined) ([raw file](https://github.com/grom72/daos/blob/871c4485d65f869bbf44043da4145c32445afc95/src/common/tests/btree.c#L1061)):*
> ```C
> 	if (argc != optind) {
> 		print_message("Unknown parameter: %s\n", argv[optind]);
> 		return -1;
> ```





_Suggestion:_
```C
		fail_msg("Unknown parameter: %s\n", argv[optind]);
```
___
*[`src/common/tests/btree.c` line 1078 at r2](https://reviewable.io/reviews/grom72/daos/79#-OILV1C4DYVFjOroubIF:-OILV1C4DYVFjOroubIG:undefined) ([raw file](https://github.com/grom72/daos/blob/871c4485d65f869bbf44043da4145c32445afc95/src/common/tests/btree.c#L1078)):*
> ```C
> 	if (rc != 0) {
> 		print_message("daos_debug_init() failed: %d\n", rc);
> 		return rc;
> ```





_Suggestion:_
```C
		fail_msg("daos_debug_init() failed: %d\n", rc);
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 2 files at r1, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: all files reviewed, 12 unresolved discussions (waiting on @janekmi)

___
*[`src/common/tests/btree.c` line 1073 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKjJH58mNrwrH803UC:-OILc34f3QIpbIenwoxY:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.c#L1073)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Nope.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 916 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILVYeN7ewWIvm3SS0j:-OILa_xKBXBA1ZWFzHzw:b-l03fo1) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L916)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

IMHO it is a string not a character.
</blockquote></details>

We expect character
___
*[`src/common/tests/btree.c` line 940 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILVn0A4BXfmLE6Jf_p:-OILahbiC0YFHTEy7pKF:b3kqx) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L940)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

.
</blockquote></details>

.
___
*[`src/common/tests/btree.c` line 1073 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILWXzwAbnQBCXlUISI:-OILbKWw5eF7F7rW8Cdj:b-ff0z8t) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L1073)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

It doesn't look right. Are you sure you need this despite the `+` at the beginning of the options string?
And how it was not necessary in the previous commit?
</blockquote></details>

`+` detects 'ma kota` as invalid parameter string but ignores `-t`- it is not an error it is unknown parameter (but with proper syntax)
___
*[`src/common/tests/btree.sh` line 34 at r1](https://reviewable.io/reviews/grom72/daos/79#-OIKhXJl29BGlZROsBGd:-OILgqPlE3G3YmK7sF4K:bb6un1j) ([raw file](https://github.com/grom72/daos/blob/6aef5a9f0befd87c552d745f5b26d8381527ec45/src/common/tests/btree.sh#L34)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Please do it in the first commit.
</blockquote></details>

Done
___
*[`src/common/tests/btree.sh` line 114 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILX7NM03SnFNT2qZGK:-OILgw5ODe6CJFVqDgTh:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L114)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

It would be nice to either separate value from option with a space for all the options or none of them.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.sh` line 134 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILXKBq9bg7qFGHJKaI:-OILgxuYE96SfVtIT0LS:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L134)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.sh` line 143 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILXLYD3DQjFCBfoRTL:-OILgyTP4Urwp4WcnAXj:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L143)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.sh` line 150 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILXMHFAESyEpE8BAe8:-OILgyku2elnTLLjnO89:b-896fix) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.sh#L150)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

.
</blockquote></details>

Done.
___
*[`src/common/tests/btree.c` line 1061 at r2](https://reviewable.io/reviews/grom72/daos/79#-OILUxRSCtZZSNrPs_Vd:-OILbteWAeCrxkwzvxGU:b-2iifyj) ([raw file](https://github.com/grom72/daos/blob/871c4485d65f869bbf44043da4145c32445afc95/src/common/tests/btree.c#L1061)):*
> ```C
> 	if (argc != optind) {
> 		print_message("Unknown parameter: %s\n", argv[optind]);
> 		return -1;
> ```

let's not mix btree.c errors and test errors
___
*[`src/common/tests/btree.c` line 1078 at r2](https://reviewable.io/reviews/grom72/daos/79#-OILV1C4DYVFjOroubIF:-OILc3o_-ZqmJ6fZz2O6:b3kqx) ([raw file](https://github.com/grom72/daos/blob/871c4485d65f869bbf44043da4145c32445afc95/src/common/tests/btree.c#L1078)):*
> ```C
> 	if (rc != 0) {
> 		print_message("daos_debug_init() failed: %d\n", rc);
> 		return rc;
> ```

.
___
*[`src/common/tests/btree.c` line 892 at r3](https://reviewable.io/reviews/grom72/daos/79#-OILKy_6-3xZqqghgIo-:-OILac6X7FHx8wvtsPyO:b-704pz) ([raw file](https://github.com/grom72/daos/blob/6eed72b325468d984e9c68456c64cb8947174875/src/common/tests/btree.c#L892)):*
> ```C
>     {"start-test", required_argument, NULL, 'S'},
>     {"reg_class", optional_argument, NULL, 'R'},
>     {"reg_memory", optional_argument, NULL, 'M'},
> ```

see `del_retain`


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: all files reviewed, 12 unresolved discussions (waiting on @janekmi)



<!-- Sent from Reviewable.io -->

janekmi reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 9 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79#-:-OILnvUW6zrk_lCCiqou:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 8 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 committed:
No space before optional option's argument

-R and -M options have optional argument.
The optional argument must be specified on the command line
without space. Otherwise, the argument is incorectly interpreted
as the next option (invalid one).

From getopt() documentation:
about option with mandatory argument:
"optstring is a string containing the legitimate option characters.
If such a character is followed by a colon, the option requires
an argument, so getopt() places a pointer to the following text
in the same argv-element, or the text of the following argv-element,
in optarg."
about option with optional argument:
"Two colons mean an option takes an optional arg; if there is text
in the current argv-element (i.e., in the same word as the option
name itself, for example, "-oarg"), then it is returned in optarg,
otherwise optarg is set to zero."

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@hpe.com>

grom72 committed:
Adapt btree_direct.c to follow btree.c

btree_direct.c and btree.c must follow the same parameters
except dyenamic registration and drain.

Unify test execution implementation.

Limit validation scope:
Skip-func-hw-test: true
Skip-func-test: true

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@hpe.com>

grom72 reviewed (commented):
> Reviewed 45 of 45 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 8 of 51 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: 9 of 52 files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->

janekmi reviewed (commented):
> Reviewed 1 of 8 files at r6, 2 of 45 files at r7, 9 of 51 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: 15 of 52 files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[`-- commits` line 42 at r7](https://reviewable.io/reviews/grom72/daos/79#-OIk3_qa16zwGbZei0_2:-OIk3_qbBJkR3dzIqKXK:bldlpbb):*
Is it right?



_Code quote:_
```
  without space. Otherwise, the argument is incorectly interpreted
  as the next option (invalid one).
```
___
*[`src/common/tests/btree.c` line 978 at r8](https://reviewable.io/reviews/grom72/daos/79#-OIk50We1en5g9Fhb_uZ:-OIk50We1en5g9Fhb_u_:b-a7aca9) ([raw file](https://github.com/grom72/daos/blob/a5d5f74aa461af2f1315bacbfba1037623d9fc25/src/common/tests/btree.c#L978)):*
> ```C
> 			break;
> 		case 'M':
> 			ik_btr_memory_register(tst_fn_val.optval);
> ```

Either `tst_fn_val` is global or you pass its value as an argument.



_Code quote:_
```C
			ik_btr_class_register(tst_fn_val.optval);
			break;
		case 'M':
			ik_btr_memory_register(tst_fn_val.optval);
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: 15 of 52 files reviewed, 2 unresolved discussions (waiting on @janekmi)

___
*[`-- commits` line 42 at r7](https://reviewable.io/reviews/grom72/daos/79#-OIk3_qa16zwGbZei0_2:-OIk6p2lElolskrMpP0z:b-lail2z):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Is it right?
</blockquote></details>

Yes. The only solution to use space is to use `'` e.g. `'-M p'`
I have updated commit message.
___
*[`src/common/tests/btree.c` line 978 at r8](https://reviewable.io/reviews/grom72/daos/79#-OIk50We1en5g9Fhb_uZ:-OIkK55k67aRgfgQv2-X:b2r3ngy) ([raw file](https://github.com/grom72/daos/blob/a5d5f74aa461af2f1315bacbfba1037623d9fc25/src/common/tests/btree.c#L978)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Either `tst_fn_val` is global or you pass its value as an argument.
</blockquote></details>

Let's fix it separately.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 43 of 51 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: all files reviewed, 2 unresolved discussions (waiting on @janekmi)



<!-- Sent from Reviewable.io -->

janekmi reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/daos/79)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @grom72)



<!-- Sent from Reviewable.io -->
